### PR TITLE
Juicify must now use the controller

### PR DESCRIPTION
**Pull request by dave-juicelabs**

### CODE REVIEW

File: pkg/restapi/client.go [hunk 1]
```diff
@@ -44,15 +68,17 @@ func (api Client) do(ctx context.Context, method string, path string, contentTyp
 	pathUrl.Scheme = "https"
 	pathUrl.Host = api.Address
 
-	response, err := api.doUrl(ctx, method, pathUrl, contentType, body)
+	response, err := api.doUrl(ctx, method, pathUrl.String(), contentType, body)
 
 	if err == nil {
-		return response, err
+		return response, nil
+	} else if err != ErrInvalidScheme {
```
Comment (ramon-juicelabs) [inline pkg/restapi/client.go:75]: This should be `!errors.Is(err, ErrInvalidScheme)`

File: cmd/juicify/app/juicify.go [hunk 2]
```diff
@@ -5,66 +5,124 @@ package app
...
+	if err != nil {
+		return err
+	}
 
-	SwapChainBuffers int `json:"swapChainBuffers,omitempty"`
+	if session.State == restapi.SessionQueued {
+		logger.Info("Session queued")
 
-	WaitForDebugger bool `json:"waitForDebugger,omitempty"`
+		ticker := time.NewTicker(5 * time.Second)
```
Comment (ykagan) [inline cmd/juicify/app/juicify.go:79]: nit: probably want to poll more frequently than this? once very second?

File: cmd/agent/app/controller.go [hunk 3]
```diff
@@ -46,133 +46,133 @@ type controllerData struct {
...
 
-					for _, session := range controllerAgent.Sessions {
-						reference, err_ := agent.getSession(session.Id)
+		for index, gpu := range gpus {
+			agent.gpuMetrics[index] = gpu.Metrics
+		}
+	})
+
+	group.GoFn("Controller Update", func(group task.Group) error {
+		ticker := time.NewTicker(5 * time.Second)
```
Comment (ykagan) [inline cmd/agent/app/controller.go:98]: nit: probably want to update more frequently than this? once very second?

File: cmd/juicify/app/juicify.go [hunk 4]
```diff
@@ -5,66 +5,124 @@ package app
...
 
-	AccessToken string `json:"accessToken,omitempty"`
+			case <-ticker.C:
+				session, err = api.GetSessionWithContext(group.Ctx(), id)
+			}
+
+			if err != nil {
+				logger.Info("Cancelling session")
+
+				api.CancelSession(id)
```
Comment (ykagan) [inline cmd/juicify/app/juicify.go:94]: can we also call this when the user ctrl-c's out of the session?

**Review by ykagan — DISMISSED** [review]

**Review by ramon-juicelabs — CHANGES_REQUESTED** [review]
Just need that error thingie below fixed as it prevents this branch from running

File: cmd/juicify/app/juicify.go (hunk 4 above)
Comment (dave-juicelabs) [reply cmd/juicify/app/juicify.go:94]: Going to put this in another PR. There are a few places where it needs to go.

**Review by ramon-juicelabs — APPROVED** [review]

**Review by ykagan — APPROVED** [review]

**Pull request closed without merging**